File: unixfs/fs-util.go
package unixfs

import (
	"context"
	"io"
	"io/fs"
	"path"
	"slices"
	"time"

	unixfs_errors "github.com/aperturerobotics/hydra/unixfs/errors"
	"github.com/pkg/errors"
)

// MaxReadFileSize is the maximum size for ReadFile operations (4GB)
const MaxReadFileSize = 4 * 1024 * 1024 * 1024

// NewReadFileSizeTooLargeError returns a standardized error for when a file is too large to be read.
func NewReadFileSizeTooLargeError(size uint64) error {
	return errors.Errorf("file size too large for ReadFile: %d bytes", size)
}

// ReaddirAllToFileInfo calls readdir and generates FileInfo objects.
// If skip is set, skips N entries.
// If limit is set, limits output to N entries.
func ReaddirAllToFileInfo(ctx context.Context, skip, limit uint64, h *FSHandle) ([]fs.FileInfo, error) {
	var children []string
	err := h.ReaddirAll(ctx, skip, func(ent FSCursorDirent) error {
		children = append(children, ent.GetName())
		if limit != 0 && len(children) >= int(limit) {
			return io.EOF
		}
		return nil
	})
	if err != nil && err != io.EOF {
		return nil, err
	}

	// lookup all children
	out := make([]fs.FileInfo, 0, len(children))
	for _, childName := range children {
		ch, err := h.Lookup(ctx, childName)
		if err == unixfs_errors.ErrNotExist {
			continue
		}
		if err != nil {
			return nil, errors.Wrapf(err, "lookup child %q", childName)
		}
		fi, err := ch.GetFileInfo(ctx)
		if err != nil {
			return nil, errors.Wrapf(err, "lookup child %q: file info", childName)
		}
		out = append(out, fi)
	}
	return out, nil
}

// ReaddirAllToFSDirEntry calls readdir and generates FSDirEntry objects.
// If skip is set, skips N entries.
// If limit is set, limits output to N entries.
func ReaddirAllToDirEntries(ctx context.Context, skip, limit uint64, h *FSHandle) ([]fs.DirEntry, error) {
	var children []FSCursorDirent
	err := h.ReaddirAll(ctx, skip, func(ent FSCursorDirent) error {
		children = append(children, ent)
		if limit != 0 && len(children) >= int(limit) {
			return io.EOF
		}
		return nil
	})
	if err != nil && err != io.EOF {
		return nil, err
	}

	// lookup all children & build directory entries
	out := make([]fs.DirEntry, 0, len(children))
	for _, child := range children {
		childName := child.GetName()
		ch, err := h.Lookup(ctx, childName)
		if err == unixfs_errors.ErrNotExist {
			continue
		}
		if err != nil {
			return nil, errors.Wrapf(err, "lookup child %q", childName)
		}
		fi, err := ch.GetFileInfo(ctx)
		if err != nil {
			return nil, errors.Wrapf(err, "lookup child %q: file info", childName)
		}
		out = append(out, NewFSDirEntry(child, fi))
	}

	return out, nil
}

// RenameWithPaths renames using two paths within a FSHandle.
func RenameWithPaths(ctx context.Context, h *FSHandle, oldPath, newPath string, ts time.Time) error {
	oldPathPts, oldPathAbsolute := SplitPath(oldPath)
	if oldPathAbsolute {
		return unixfs_errors.ErrAbsolutePath
	}

	newPathPts, newPathAbsolute := SplitPath(newPath)
	if newPathAbsolute {
		return unixfs_errors.ErrAbsolutePath
	}

	if slices.Equal(oldPathPts, newPathPts) {
		return nil
	}

	if len(newPath) == 0 || len(oldPath) == 0 {
		return unixfs_errors.ErrEmptyPath
	}

	oldHandle, _, err := h.LookupPathPts(ctx, oldPathPts)
	if err != nil {
		if oldHandle != nil {
			oldHandle.Release()
		}
		return err
	}
	defer oldHandle.Release()

	parentPathPts := newPathPts[:len(newPath)-1]
	destName := newPathPts[len(newPath)-1]
	nextParent, _, err := h.LookupPathPts(ctx, parentPathPts)
	if err != nil {
		if nextParent != nil {
			nextParent.Release()
		}
		return err
	}
	defer nextParent.Release()

	return oldHandle.Rename(ctx, nextParent, destName, ts)
}

// StatWithPath calls Stat on a path in a FSHandle.
// Note: this will traverse Symbolic links.
// TODO: LStat (don't traverse symbolic links.)
func StatWithPath(ctx context.Context, h *FSHandle, name string) (fs.FileInfo, error) {
	name = path.Clean(name)
	if name == "." || name == "/" || name == "" {
		return h.GetFileInfo(ctx)
	}

	fh, _, err := h.LookupPath(ctx, name)
	if err != nil {
		if fh != nil {
			fh.Release()
		}
		return nil, err
	}
	defer fh.Release()

	return fh.GetFileInfo(ctx)
}

// RemoveAllWithPath calls Remove on the given path.
// Returns ErrNotExist if the path didn't exist.
func RemoveAllWithPath(ctx context.Context, h *FSHandle, filepath string, ts time.Time) error {
	filepath = path.Clean(filepath)
	filedir, filename := path.Split(filepath)
	dirHandle, _, err := h.LookupPath(ctx, filedir)
	if err != nil {
		if dirHandle != nil {
			dirHandle.Release()
		}
		return err
	}
	defer dirHandle.Release()

	return dirHandle.Remove(ctx, []string{filename}, ts)
}

// ChmodWithPath calls Chmod on the given path.
func ChmodWithPath(ctx context.Context, h *FSHandle, filepath string, mode fs.FileMode, ts time.Time) error {
	ch, _, err := h.LookupPath(ctx, filepath)
	if err != nil {
		if ch != nil {
			ch.Release()
		}
		return err
	}
	defer ch.Release()

	info, err := ch.GetFileInfo(ctx)
	if err != nil {
		return err
	}

	oldType := info.Mode() & fs.ModeType
	setType := mode & fs.ModeType
	if oldType != setType {
		return errors.New("TODO chmod: change node type")
	}

	oldPerms := info.Mode() & fs.ModePerm
	setPerms := mode & fs.ModePerm
	if oldPerms != setPerms {
		err = ch.SetPermissions(ctx, setPerms, ts)
		if err != nil {
			return err
		}
	}
	return nil
}

// SetModTimestampWithPath changes the modification timestamp on the given path.
// mtime is the modification time to use.
func SetModTimestampWithPath(ctx context.Context, h *FSHandle, filepath string, mtime time.Time) error {
	ch, _, err := h.LookupPath(ctx, filepath)
	if err != nil {
		if ch != nil {
			ch.Release()
		}
		return err
	}
	defer ch.Release()

	return ch.SetModTimestamp(ctx, mtime)
}

// ReadFile reads the named file and returns the contents.
// A successful call returns err == nil, not err == EOF.
// Because ReadFile reads the whole file, it does not treat an EOF from Read
// as an error to be reported.
func ReadFile(ctx context.Context, h *FSHandle) ([]byte, error) {
	var size int64
	if info, err := h.GetFileInfo(ctx); err == nil {
		size = info.Size()
	}
	if size == 0 {
		return nil, nil
	}

	// If a file claims a small size, read at least 512 bytes.
	if size < 512 {
		size = 512
	} else if size > MaxReadFileSize {
		return nil, NewReadFileSizeTooLargeError(uint64(size))
	} else {
		size++ // one byte for final read at EOF
	}

	data := make([]byte, 0, size)
	for {
		if len(data) >= cap(data) {
			d := append(data[:cap(data)], 0)
			data = d[:len(data)]
		}
		n, err := h.ReadAt(ctx, int64(len(data)), data[len(data):cap(data)])
		data = data[:len(data)+int(n)]
		if err != nil {
			if err == io.EOF {
				err = nil
			}
			return data, err
		}
	}
}

// WriteFile writes data to the filesystem handle, which must be a file.
//
// Since WriteFile requires multiple system calls to complete, a failure mid-operation
// can leave the file in a partially written state.
func WriteFile(ctx context.Context, fsh *FSHandle, data []byte, ts time.Time) error {
	optimalWriteSize, err := fsh.GetOptimalWriteSize(ctx)
	if err != nil {
		return err
	}
	if optimalWriteSize == 0 {
		// copy the const here to avoid the blob import
		optimalWriteSize = 2048 * 125 // 256KB - blob.DefChunkingMinSize
	}

	// Truncate the file
	err = fsh.Truncate(ctx, 0, ts)
	if err != nil {
		return err
	}

	// Write data in chunks
	for offset := int64(0); offset < int64(len(data)); offset += int64(optimalWriteSize) {
		end := min(offset+int64(optimalWriteSize), int64(len(data)))
		chunk := data[offset:end]

		err = fsh.WriteAt(ctx, offset, chunk, ts)
		if err != nil {
			return err
		}
	}

	return nil
}

File: unixfs/fs-util.ts
import type { FSCursorDirent } from './fs-cursor.js'
import type { FSHandle } from './fs-handle.js'
import {
  ErrAbsolutePath,
  ErrEmptyPath,
  isUnixFSError,
} from './errors/errors.js'
import { UnixFSErrorType } from './errors/errors.pb.js'
import { splitPath } from './path.js'

// MAX_READ_FILE_SIZE is the maximum size for readFile operations (4GB).
export const MAX_READ_FILE_SIZE = 4n * 1024n * 1024n * 1024n

// newReadFileSizeTooLargeError returns an error for when a file is too large.
export function newReadFileSizeTooLargeError(size: bigint): Error {
  return new Error(`file size too large for readFile: ${size} bytes`)
}

// lookupPathOrThrow is a helper that calls lookupPath and throws on error,
// releasing the partial handle. Returns the handle on success.
async function lookupPathOrThrow(
  signal: AbortSignal,
  h: FSHandle,
  path: string,
): Promise<FSHandle> {
  const result = await h.lookupPath(signal, path)
  if (result.error) {
    result.handle.release()
    throw result.error
  }
  return result.handle
}

// lookupPathPtsOrThrow is a helper that calls lookupPathPts and throws on error,
// releasing the partial handle. Returns the handle on success.
async function lookupPathPtsOrThrow(
  signal: AbortSignal,
  h: FSHandle,
  parts: string[],
): Promise<FSHandle> {
  const result = await h.lookupPathPts(signal, parts)
  if (result.error) {
    result.handle.release()
    throw result.error
  }
  return result.handle
}

// readFile reads the named file and returns the contents.
export async function readFile(
  signal: AbortSignal,
  h: FSHandle,
): Promise<Uint8Array> {
  let size = 0n
  try {
    const info = await h.getFileInfo(signal)
    size = info.size
  } catch {
    // ignore
  }
  if (size === 0n) {
    return new Uint8Array(0)
  }

  // If a file claims a small size, read at least 512 bytes.
  if (size < 512n) {
    size = 512n
  } else if (size > MAX_READ_FILE_SIZE) {
    throw newReadFileSizeTooLargeError(size)
  } else {
    size++ // one byte for final read at EOF
  }

  const chunks: Uint8Array[] = []
  let offset = 0n
  let totalRead = 0
  for (;;) {
    const remaining = Number(size) - totalRead
    if (remaining <= 0) {
      size = size + 512n
    }

    const chunkSize = Math.min(Number(size) - totalRead, 65536)
    if (chunkSize <= 0) break
    const buf = new Uint8Array(chunkSize)
    let n: bigint
    try {
      n = await h.readAt(signal, offset, buf)
    } catch (err) {
      if (isUnixFSError(err, UnixFSErrorType.EOF)) {
        break
      }
      throw err
    }
    if (n === 0n) {
      break
    }
    chunks.push(buf.subarray(0, Number(n)))
    totalRead += Number(n)
    offset += n
    if (n < BigInt(chunkSize)) {
      break
    }
  }

  // Concatenate chunks.
  const result = new Uint8Array(totalRead)
  let pos = 0
  for (const chunk of chunks) {
    result.set(chunk, pos)
    pos += chunk.length
  }
  return result
}

// DEFAULT_CHUNK_SIZE is the default optimal write size (256KB).
const DEFAULT_CHUNK_SIZE = 2048n * 125n

// writeFile writes data to the filesystem handle, which must be a file.
export async function writeFile(
  signal: AbortSignal,
  h: FSHandle,
  data: Uint8Array,
  ts: Date,
): Promise<void> {
  let optimalWriteSize = await h.getOptimalWriteSize(signal)
  if (optimalWriteSize === 0n) {
    optimalWriteSize = DEFAULT_CHUNK_SIZE
  }

  // Truncate the file.
  await h.truncate(signal, 0n, ts)

  // Write data in chunks.
  const dataLen = BigInt(data.length)
  for (let offset = 0n; offset < dataLen; offset += optimalWriteSize) {
    const end = offset + optimalWriteSize < dataLen
      ? offset + optimalWriteSize
      : dataLen
    const chunk = data.subarray(Number(offset), Number(end))
    await h.writeAt(signal, offset, chunk, ts)
  }
}

// statWithPath calls stat on a path in a FSHandle.
// This will traverse symbolic links.
export async function statWithPath(
  signal: AbortSignal,
  h: FSHandle,
  name: string,
): Promise<{
  name: string
  size: bigint
  mode: number
  modTime: Date
  isDir: boolean
}> {
  name = cleanPath(name)
  if (name === '.' || name === '/' || name === '') {
    return h.getFileInfo(signal)
  }

  const fh = await lookupPathOrThrow(signal, h, name)
  try {
    return await fh.getFileInfo(signal)
  } finally {
    fh.release()
  }
}

// lstatWithPath calls lstat on a path in a FSHandle.
// Unlike statWithPath, the final path component is not followed if it is a symlink.
export async function lstatWithPath(
  signal: AbortSignal,
  h: FSHandle,
  name: string,
): Promise<{
  name: string
  size: bigint
  mode: number
  modTime: Date
  isDir: boolean
}> {
  name = cleanPath(name)
  if (name === '.' || name === '/' || name === '') {
    return h.getFileInfo(signal)
  }

  const lastSlash = name.lastIndexOf('/')
  const dir = lastSlash >= 0 ? name.substring(0, lastSlash) : ''
  const base = lastSlash >= 0 ? name.substring(lastSlash + 1) : name

  // Look up the parent directory (follows symlinks for intermediate components).
  let parentHandle: FSHandle
  if (dir === '' || dir === '.' || dir === '/') {
    parentHandle = await h.clone(signal)
  } else {
    parentHandle = await lookupPathOrThrow(signal, h, dir)
  }

  try {
    // Single lookup on the final component does NOT follow symlinks.
    const childHandle = await parentHandle.lookup(signal, base)
    try {
      return await childHandle.getFileInfo(signal)
    } finally {
      childHandle.release()
    }
  } finally {
    parentHandle.release()
  }
}

// renameWithPaths renames using two paths within a FSHandle.
export async function renameWithPaths(
  signal: AbortSignal,
  h: FSHandle,
  oldPath: string,
  newPath: string,
  ts: Date,
): Promise<void> {
  const { parts: oldPathPts, isAbsolute: oldAbs } = splitPath(oldPath)
  if (oldAbs) {
    throw ErrAbsolutePath
  }

  const { parts: newPathPts, isAbsolute: newAbs } = splitPath(newPath)
  if (newAbs) {
    throw ErrAbsolutePath
  }

  if (arraysEqual(oldPathPts, newPathPts)) {
    return
  }

  if (!newPath || !oldPath) {
    throw ErrEmptyPath
  }

  const oldHandle = await lookupPathPtsOrThrow(signal, h, oldPathPts)
  try {
    const parentPathPts = newPathPts.slice(0, -1)
    const destName = newPathPts[newPathPts.length - 1]
    const nextParent = await lookupPathPtsOrThrow(signal, h, parentPathPts)
    try {
      await oldHandle.rename(signal, nextParent, destName, ts)
    } finally {
      nextParent.release()
    }
  } finally {
    oldHandle.release()
  }
}

// removeAllWithPath calls remove on the given path.
export async function removeAllWithPath(
  signal: AbortSignal,
  h: FSHandle,
  filepath: string,
  ts: Date,
): Promise<void> {
  filepath = cleanPath(filepath)
  const lastSlash = filepath.lastIndexOf('/')
  const filedir = lastSlash >= 0 ? filepath.substring(0, lastSlash) : ''
  const filename =
    lastSlash >= 0 ? filepath.substring(lastSlash + 1) : filepath

  const dirHandle = await lookupPathOrThrow(signal, h, filedir || '.')
  try {
    await dirHandle.remove(signal, [filename], ts)
  } finally {
    dirHandle.release()
  }
}

// chmodWithPath calls chmod on the given path.
export async function chmodWithPath(
  signal: AbortSignal,
  h: FSHandle,
  filepath: string,
  mode: number,
  ts: Date,
): Promise<void> {
  const ch = await lookupPathOrThrow(signal, h, filepath)
  try {
    const info = await ch.getFileInfo(signal)
    const oldPerms = info.mode & 0o777
    const setPerms = mode & 0o777
    if (oldPerms !== setPerms) {
      await ch.setPermissions(signal, setPerms, ts)
    }
  } finally {
    ch.release()
  }
}

// setModTimestampWithPath changes the modification timestamp on the given path.
export async function setModTimestampWithPath(
  signal: AbortSignal,
  h: FSHandle,
  filepath: string,
  mtime: Date,
): Promise<void> {
  const ch = await lookupPathOrThrow(signal, h, filepath)
  try {
    await ch.setModTimestamp(signal, mtime)
  } finally {
    ch.release()
  }
}

// readdirAllToEntries reads all directory entries and returns file info.
export async function readdirAllToEntries(
  signal: AbortSignal,
  h: FSHandle,
  skip: bigint,
  limit: number,
): Promise<
  Array<{
    name: string
    size: bigint
    mode: number
    modTime: Date
    isDir: boolean
  }>
> {
  const children: string[] = []
  try {
    await h.readdirAll(signal, skip, (ent: FSCursorDirent) => {
      children.push(ent.getName())
      if (limit > 0 && children.length >= limit) {
        throw new EofSignal()
      }
    })
  } catch (e) {
    if (!(e instanceof EofSignal)) {
      throw e
    }
  }

  const out: Array<{
    name: string
    size: bigint
    mode: number
    modTime: Date
    isDir: boolean
  }> = []
  for (const childName of children) {
    let ch: FSHandle
    try {
      ch = await h.lookup(signal, childName)
    } catch (e) {
      if (isUnixFSError(e, UnixFSErrorType.NOT_EXIST)) {
        continue
      }
      throw e
    }
    try {
      const fi = await ch.getFileInfo(signal)
      out.push(fi)
    } finally {
      ch.release()
    }
  }
  return out
}

// EofSignal is used to signal EOF in readdirAll callbacks.
class EofSignal extends Error {
  constructor() {
    super('EOF')
  }
}

// arraysEqual checks if two string arrays are equal.
function arraysEqual(a: string[], b: string[]): boolean {
  if (a.length !== b.length) return false
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false
  }
  return true
}

// cleanPath is a basic path.Clean equivalent for local use.
function cleanPath(p: string): string {
  if (p === '' || p === '.') return '.'
  const parts: string[] = []
  const isAbs = p[0] === '/'
  const segs = p.split('/')
  for (const seg of segs) {
    if (seg === '' || seg === '.') continue
    if (seg === '..') {
      if (parts.length > 0 && parts[parts.length - 1] !== '..') {
        parts.pop()
      } else if (!isAbs) {
        parts.push('..')
      }
    } else {
      parts.push(seg)
    }
  }
  let result = parts.join('/')
  if (isAbs) {
    result = '/' + result
  }
  return result || (isAbs ? '/' : '.')
}
